pennylane/ops/functions/matrix.py

```diff
@@ -134,13 +134,14 @@ def _matrix(tape, wire_order=None):
     interface = qml.math._multi_dispatch(params)
 
     wire_order = wire_order or tape.wires
-    n_wires = len(wire_order)
 
     # initialize the unitary matrix
-    unitary_matrix = qml.math.eye(2**n_wires, like=interface)
+    result = qml.math.eye(2 ** len(wire_order), like=interface)
 
     for op in tape.operations:
         U = matrix(op, wire_order=wire_order)
-        unitary_matrix = qml.math.tensordot(U, unitary_matrix, axes=[[-1], [-2]])
+        # Coerce the matrices U and result and use matrix multiplication. Broadcasted axes
+        # are handled correctly automatically by ``matmul`` (See e.g. NumPy documentation)
+        result = qml.math.matmul(*qml.math.coerce([U, result], like=interface), like=interface)

```

Comment:
:tada: :tada: :tada:

Reply:
For future reference: An alternative here would be to make a `qml.math.matmul` modification analogous to `qml.math.tensordot`, that always calls coerce before calling the autoray `matmul` dispatcher. However, I thought this is overkill here.